src/xpk/core/ray.py

```diff
@@ -0,0 +1,203 @@
+            cloud.google.com/gke-tpu-topology: {topology}
+"""
+
+
+def install_ray_cluster(args, system) -> int:
+  """Install a RayCluster on the cluster
+
+  Args:
+    args: user provided arguments for running the command.
+    system: system characteristics.
+
+  Returns:
+    0 if successful and 1 otherwise.
+  """
+
+  delete_ray_cluster(args)
+
+  label = 'cloud.google.com/gke-nodepool=default-pool'
+  available_head_cpu, available_head_mem = generate_available_resources(
+      label, args, 0.5
+  )
+
+  label = f'cloud.google.com/gke-tpu-accelerator={system.gke_accelerator}'
+  available_worker_cpu, available_worker_mem = generate_available_resources(
+      label, args, 0.9
+  )
+
+  yml_string = ray_cluster_crd_yaml.format(
+      accelerator=system.gke_accelerator,
+      topology=system.topology,
+      chips_per_vm=system.chips_per_vm,
+      num_hosts=system.vms_per_slice,
+      replicas=args.num_slices,
+      version=args.ray_version,
+      worker_cpu=available_worker_cpu,
+      worker_mem=available_worker_mem,
+      head_cpu=available_head_cpu,
+      head_mem=available_head_mem,
+  )
+
+  tmp = write_tmp_file(yml_string)
+  command = f'kubectl apply -f {str(tmp.file.name)}'
+  task = 'Applying RayCluster'
+  retry_attempts = 1
+  return_code = run_command_with_updates_retry(
+      command, task, args, num_retry_attempts=retry_attempts
+  )
+  if return_code != 0:
+    xpk_print(f'{task} not successful.')
+    xpk_exit(return_code)
+  return return_code
+
+
+def delete_ray_cluster(args) -> None:
+  """Delete all RayClusters on the cluster
+
+  Args:
+    args: user provided arguments for running the command.
+
+  Returns:
+    None
+  """
+
+  command = 'kubectl delete rayclusters -n ray --all'
+  task = 'Deleting old RayCluster'
+  retry_attempts = 1
+  return_code = run_command_with_updates_retry(
+      command, task, args, num_retry_attempts=retry_attempts
+  )
+
+  if return_code != 0:
+    xpk_print(f'{task} not successful.')
+    xpk_exit(return_code)
+
+  return
+
+
+def generate_available_resources(label, args, percent) -> tuple:
+  """Generate the available resources for the nodes that match the given label
+
+  Args:
+    label: the label used to match the appropriate nodes
+    args: user provided arguments for running the command
+    percent: the percent of the available resources to use
+
+  Returns:
+    A tuple with the available cpu and memory
+  """
+
+  command = (
+      f"kubectl get nodes -l {label} -o jsonpath='{{.items[0].metadata.name}}'"
+  )
+  task = f'Getting nodes with label {label}'
+  _, node_name = run_command_for_value(command, task, args)
+
+  command = (
+      f"kubectl get node {node_name} -o jsonpath='{{.status.allocatable.cpu}}'"
+  )
+  task = 'Fetching available CPU on node'
+  _, available_cpu = run_command_for_value(command, task, args)
+  match = re.match(r'(\d+)([a-zA-Z]+)', available_cpu)
+  if not match:
+    xpk_exit(1)
```

Comment:
please add xpk_print with explanation why exiting (generating available resources failed due to ...)

Reply:
Done.